File: blackjack/scala/Cards.scala
/*
 * Main.scala
 *
 * To change this template, choose Tools | Template Manager
 * and open the template in the editor.
 */

package com.draga.blackjack.Cards

sealed trait suit
final case object Hearts extends suit
final case object Diamonds extends suit
final case object Clubs extends suit
final case object Spades extends suit

sealed trait value
final case object Ace extends value
final case object Two extends value
final case object Three extends value
final case object Four extends value
final case object Five extends value
final case object Six extends value
final case object Seven extends value
final case object Eight extends value
final case object Nine extends value
final case object Ten extends value
final case object Jack extends value
final case object Queen extends value
final case object King extends value

case class Card (s: suit, v: value) {
  override def toString = v + " of " + s
}

object Cards {
  val suits = Array(Hearts, Diamonds, Clubs, Spades)
  val values = Array(Ace, Two, Three, Four, Five, Six, Seven,
                             Eight, Nine, Ten, Jack, Queen, King)
  val oneDeck: Array[Card] = for (s <- suits; v <- values) yield Card(s,v)
}

object Shoe {
  // XXX should be configurable
  val decksInShoe = 6
  // XXX should use List (see draw!), but need to rewrite shuffle
  private var cards : Array[Card] = Array()
  private val rand = new util.Random();

  //    Shuffle due to John Lees-Miller at
  //      http://jdleesmiller.blogspot.com/2008/12/shuffles-surprises-and-scala.html

  private def swap[T](xs: Array[T], i: Int, j: Int) = {
    val t = xs(i)
    xs(i) = xs(j)
    xs(j) = t
  }

  private def fisherYatesShuffle[T](xs: Array[T]) = {
    for (i <- xs.indices.reverse) swap(xs, i, rand.nextInt(i + 1))
  }

  def shuffle : Unit = {
    var i : Int = 0
    cards = Array()
    for (i <- 1 to decksInShoe) {
      cards ++= Cards.oneDeck
    }
    fisherYatesShuffle(cards)
  }

  def draw : Card = {
    if (cards.isEmpty) shuffle

    val c = cards.head
    cards = cards.drop(1)
    c
  }
}

File: blackjack/c/cards.c
#include <sys/types.h>
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>

#include "cards.h"

const int decksinshoe = 6;
const int cardsindeck = 52;
const int cardsinshoe = 6 * 52;
static int currcard;

const char *suit_names[] = {"Hearts", "Diamonds", "Clubs", "Spades"};
const char *value_names[] = {"Ace", "Two", "Three", "Four", "Five", "Six", "Seven",
			     "Eight", "Nine", "Ten", "Jack", "Queen", "King"};

static struct card *onedeck = NULL;
static struct card *shoe = NULL;

static void knuth_shuffle_cards (void);

/* uses a static buffer; not re-entrant */
static char cname[17];

const char *
card_name (card c) {
  snprintf(cname, sizeof(cname), "%s of %s", value_names[c.value], suit_names[c.suit]);
  return cname;
}

static void
init_onedeck (void) {
  enum suit s;
  enum value v;
  int i = 0;

  /* catch attempts to reinit onedeck */
  assert(onedeck == NULL);

  /* never freed; generated once */
  onedeck = malloc(cardsindeck * sizeof(struct card));
  for (s=HEARTS; s<=SPADES; s++)
    for (v=ACE; v<=KING; v++) {
      onedeck[i].suit = s;
      onedeck[i].value = v;
      i++;
    }
}

card
draw (void) {
  int i, j, k;

  if (shoe == NULL) {
    shoe = malloc(cardsinshoe * sizeof(card));
    srand48(time(NULL));
    currcard = cardsinshoe;

    /* only need this once; afterward, we shuffle the same cards again and again */
    if (onedeck == NULL)
      init_onedeck();
    for (i=0, k=0; i<decksinshoe; i++)
      for (j=0; j<cardsindeck; j++, k++)
	shoe[k] = onedeck[j];
  }

  if (currcard == cardsinshoe) {
    printf("Refilling shoe with %d decks\n", decksinshoe);
    knuth_shuffle_cards();
    currcard = 0;
  }

  return shoe[currcard++];
}

/* adapted from scheme version in ../scheme, see there for attribution */
void
knuth_shuffle_cards (void) {
  int i, r;
  struct card t;

  for (i=cardsinshoe; i>0; i--) {
    r = lrand48() % cardsinshoe;
    t = shoe[r];
    shoe[r] = shoe[i];
    shoe[i] = t;
  }
}

cardlist
clist_new (void) {
  cardlist cs = {NULL, NULL};
  return cs;
}

void
clist_add (cardlist *cs, card c) {
  card *cn;

  assert(cs != NULL);
  cn = malloc(sizeof(card)); /* freed in clist_free */
  assert(cn != NULL);

  *cn = c;

  if (cs->first == NULL) {
    cs->first = cs->last = cn;
  } else {
    cs->last->next = cn;
    cn->next = NULL;
    cs->last = cn;
  }
}

int
clist_length (cardlist cs) {
  int n = 0;
  card *c;

  for (c = cs.first; c != NULL; c = c->next)
    n++;

  return n;
}

void
clist_clear (cardlist *cs) {
  card *c;
  for (c = cs->first; c!=NULL; c = c->next)
    free(c);

  cs->first = cs->last = NULL;
}
